src/resources/FreeRun.ts

```diff
@@ -0,0 +1,272 @@
+import {
+  cliExecute,
+  mallPrice,
+  myTurncount,
+  restoreMp,
+  retrieveItem,
+  visitUrl,
+} from "kolmafia";
+
+import { Macro } from "../combat";
+import {
+  ensureEffect,
+  getFoldGroup,
+  getSongCount,
+  getSongLimit,
+  have,
+} from "../lib";
+import { Requirement } from "../maximize";
+import { get } from "../property";
+import { $effect, $familiar, $item, $items, $skill } from "../template-string";
+import {
+  ActionSource,
+  findActionSource,
+  FindActionSourceOptions,
+} from "./action";
+import * as Bandersnatch from "./2009/Bandersnatch";
+import * as AsdonMartin from "./2017/AsdonMartin";
+
+export type AdventureOptions = {
+  equipmentRequirements?: () => Requirement;
+  preparation?: () => boolean;
+  location?: () => Location;
+  macro?: () => Macro;
+  familiar?: () => Familiar;
+  available?: () => boolean;
+};
+
+// Free unlimited source every 30 turns
+// Does not work on special monsters so needs a backup, see tryFindFreeRun
+const asdonMartinSource: ActionSource = new ActionSource(
+  $skill`Asdon Martin: Spring-Loaded Front Bumper`,
+  () => {
+    if (!AsdonMartin.have()) return 0;
+    const banishes = get("banishedMonsters").split(":");
+    const bumperIndex = banishes
+      .map((string) => string.toLowerCase())
+      .indexOf("spring-loaded front bumper");
+    if (bumperIndex === -1) return 1;
+    return myTurncount() - parseInt(banishes[bumperIndex + 1]) > 30 ? 1 : 0;
+  },
+  Macro.skill($skill`Asdon Martin: Spring-Loaded Front Bumper`),
+  {
+    preparation: () => AsdonMartin.fillTo(50),
+  }
+);
+
+const freeRunSources: ActionSource[] = [
+  // Free limited sources
+  new ActionSource(
+    $familiar`Frumious Bandersnatch`,
+    () =>
+      have($familiar`Frumious Bandersnatch`) &&
+      (have($effect`Ode to Booze`) || getSongCount() < getSongLimit()) &&
+      Bandersnatch.getRemainingRunaways() > 0
+        ? 0
```

Comment:
whoops didn't implement bander or boots potential yet

Reply:
this is probably the trickiest part conceptually as I'm sure you know, unless you want to maximize for familiar weight every turn. the way I've done it in the past is to have a daily pref for when we've tried and failed, but I don't know if that is generic enough here. Maybe it is.